### PR TITLE
Saving/loading workspaces via sdk

### PR DESCRIPTION
**Pull request by swheaton**
## What changes are proposed in this pull request?

- Move `fiftyone.core.spaces` to `fiftyone.core.odm.workspace`. They were documents so belonged in `odm` to begin with I think. `fo.Space` and `fo.Panel` are still available as before (and have unchanged class definition except for module path.
- Create `fiftyone.core.odm.workspace.WorkspaceDocument` class that has a `Space` as a top level child, as well as the other metadata like description, color, created_at, etc. This is what's saved in DB.
- Copy most of the saved view implementation and paste as "workspace"

Not done: export/import/snapshot spaces (tracked in separate ticket)

## How is this patch tested? If it is not, please explain why.

Added new unit tests like saved view but for workspaces

Test yourself like this:

```python
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")

embeddings_panel = fo.Panel(
    type="Embeddings",
    state=dict(
        brainResult="img_viz",
        colorByField="metadata.size_bytes"
    ),
)
workspace = fo.Space(children=[embeddings_panel])

workspace_name = "embeddings-workspace"
description = "Show embeddings only"
dataset.save_workspace(
    workspace_name,
    workspace,
    description=description
)
assert dataset.has_workspace(workspace_name)
info = dataset.get_workspace_info(workspace_name)
assert info["description"] == description
```
New session
```python
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")

embeddings_panel = fo.Panel(
    type="Embeddings",
    state=dict(brainResult="img_viz", colorByField="metadata.size_bytes"),
)
workspace = fo.Space(children=[embeddings_panel])
workspace_name = "embeddings-workspace"
dataset.save_workspace(workspace_name, workspace)

# Some time later ... load the workspace
loaded_workspace = dataset.load_workspace(workspace_name)
assert workspace == loaded_workspace

# Launch app with the loaded workspace!
session = fo.launch_app(dataset, spaces=loaded_workspace)

# Or set via session later on
session.spaces = loaded_workspace
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

```text
Added ability to save and load :ref:`app spaces layouts <app-spaces>` as :ref:`workspaces <app-saving-spaces-python>` in the SDK.
```

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to export and import workspaces alongside datasets, enhancing data management capabilities.
	- Added comprehensive workspace management within datasets, including saving, loading, updating, and deleting workspaces.
	- Implemented new test cases for import/export consistency of workspaces and their management within datasets.
- **Refactor**
	- Improved consistency between referenced documents and dataset IDs in the database operations.
- **Documentation**
	- Updated the user guide with instructions for saving, loading, and managing workspaces in Python using the `fiftyone` library.

### CODE REVIEW

File: fiftyone/core/dataset.py
Comment (swheaton) [inline fiftyone/core/dataset.py:3685]: upgrade error message like I previously did with dataset slug clash

File: fiftyone/core/dataset.py
Comment (swheaton) [reply fiftyone/core/dataset.py:3694]: @coderabbitai no this is correct, this block of code is in the function `_validate_saved_view_name` which is for saved views. There is a similar function `_validate_workspace_name` on line 3964 that has the version for workspace name validation.

File: fiftyone/core/odm/database.py
Comment (brimoor) [inline fiftyone/core/odm/database.py:861]: @swheaton there's a few instances of "saved views" nomenclature in this method that need generalizing

Comment (swheaton) [reply fiftyone/core/odm/database.py:861]: dang it thought i got em all

File: fiftyone/core/odm/database.py
Comment (swheaton) [inline fiftyone/core/odm/database.py:848]:
```suggestion
    ids_from_dataset_list = dataset_dict.get(field_name, [])
```

File: fiftyone/core/odm/database.py
Comment (swheaton) [inline fiftyone/core/odm/database.py:872]:
```suggestion
        _logger.info(
            "Purging %d bad %s ID(s) %s from dataset",
            num_bad_ids,
            field_name,
            bad_ids,
        )
```

File: docs/source/user_guide/app.rst
Comment (brimoor) [inline docs/source/user_guide/app.rst:1549]: @swheaton @benjaminpkane in the case of saved views, the App "knows" whether the current view is named:

```py
view = dataset.load_saved_view(name)
assert view.name == name

session.view = view
session.refresh()

# The session "knows" its a saved view, also reflected in the URL via `?view=name`
assert session.view.name == name
```

This is relevant in many contexts, eg you might want to write an operator to rename the current view and thus `ctx.view` needs to "know" that the current view is named (recently supported by @imanjra in https://github.com/voxel51/fiftyone/pull/4200).

I expect the same will be true for saved workspaces:

```py
workspace = dataset.load_workspace(name)
assert workspace.name == name

session.spaces = workspace
session.refresh()

# The session "knows" its a saved workspace, also reflected in the URL via `?workspace=name`
session.spaces.name == name
```

@swheaton was initially going to propose that we have a `Workspace(Space)` class that adds a `name` property, but I suppose `Space` also needs the `name` property because this would be a natural thing to do:

```py
# Unsaved workspace
assert session.spaces.name is None
```

@benjaminpkane @imanjra for the idea of the App knowing `?workspace=name`

Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: ok I don't see an issue with that.

**Review by brimoor — COMMENTED** [review]
@swheaton really appreciate the thorough treatment here! 💪 

I believe there's just one additional place that needs treatment: the `[Legacy]FiftyOneDataset[Importer|Exporter]` classes should also include workspaces:
- https://github.com/voxel51/fiftyone/blob/develop/fiftyone/utils/data/exporters.py#L1855-L1859
- https://github.com/voxel51/fiftyone/blob/33e1d9ee6e488caf4c0c4d23aba8c8d2cf493d2c/fiftyone/utils/data/exporters.py#L2175-L2178
- https://github.com/voxel51/fiftyone/blob/33e1d9ee6e488caf4c0c4d23aba8c8d2cf493d2c/fiftyone/utils/data/importers.py#L1583-L1590
- https://github.com/voxel51/fiftyone/blob/33e1d9ee6e488caf4c0c4d23aba8c8d2cf493d2c/fiftyone/utils/data/importers.py#L1944-L1949
- https://github.com/voxel51/fiftyone/blob/33e1d9ee6e488caf4c0c4d23aba8c8d2cf493d2c/tests/unittests/import_export_tests.py#L2926-L2950

File: docs/source/user_guide/app.rst
Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: Yeah I had considered deprecating `spaces` and adding an alias argument `workspace`, but decided against it, for same reasons

**Comment by swheaton** [comment]
> @swheaton really appreciate the thorough treatment here! 💪
> 
> I believe there's just one additional place that needs treatment: the `[Legacy]FiftyOneDataset[Importer|Exporter]` classes should also include workspaces:

Yes we separated that into a different ticket for scheduling purposes (import/export not considered strictly necessary for the feature to be valid). I can add it on to this PR if I find some time.

File: docs/source/user_guide/app.rst
Comment (imanjra) [reply docs/source/user_guide/app.rst:1549]: Yah, the `Space` will need to have an optional name property to preserve the name in the session.spaces. Alternatively, `session.spaces` can be Union of `Workspace` and `Space` and workspace serialization could have the name property. Either would be fine as long we can preserve the `space.name` (at least for root space) when sending down serialized session.spaces state

Comment (benjaminpkane) [reply docs/source/user_guide/app.rst:1549]: I think a union of `Workspace` and `Space` makes sense for `session.spaces`. If not, `Spaces` (which are nest-able) has an somewhat unnecessary/misleading attribute

Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: This is true. If that's the case, I may need some pointers on implementation.

**Comment by swheaton** [comment]
Ok import/export is complete. I need to hop over to API repo to finish it for snapshots

File: docs/source/user_guide/app.rst
Comment (brimoor) [reply docs/source/user_guide/app.rst:1549]: I'm thinking we should just stick with the `Space` class with a nullable `name` property.

If we're going to be strict about backwards-compatibility, we need to still support this:

```py
session.spaces = fo.Space(...)
```

Together with my desire for `session.spaces.name` to always exist going forward, this means we have to add the `name` property to `Space` regardless.

Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: @benjaminpkane I have a solution to above that I'm ready to push, passes tests, etc. Can you give me a test that needs to be true in order for session.spaces serialization in app routes to work with name property?

> Either would be fine as long we can preserve the space.name (at least for root space) when sending down serialized session.spaces state

Comment (benjaminpkane) [reply docs/source/user_guide/app.rst:1549]: The name should be accessible in the serialize spaces via the [StateDescription](https://github.com/voxel51/fiftyone/blob/develop/fiftyone/core/state.py)

Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: Ok I'm thinking it should work as long as, say, `from_json(to_json())` round trip works and has name field, which it does.
Is there a typescript definition of statedescription i need to be concerned with?

Comment (benjaminpkane) [reply docs/source/user_guide/app.rst:1549]: Ibrahim or I can follow up with typescript addition(s)

Comment (benjaminpkane) [reply docs/source/user_guide/app.rst:1549]: > @benjaminpkane I have a solution to above that I'm ready to push, passes tests, etc.

Once you push, I can start work on the `workspace` URL parameter in earnest

Comment (swheaton) [reply docs/source/user_guide/app.rst:1549]: sorry didnt realize you were blocked on me for anything. it's pushed now

File: fiftyone/core/dataset.py
Comment (swheaton) [reply fiftyone/core/dataset.py:3706]: @coderabbitai this is a duplicate comment to one you made two days ago that I resolved. Here is the quoted comment.

> no this is correct, this block of code is in the function _validate_saved_view_name which is for saved views. There is a similar function _validate_workspace_name on line 3964 that has the version for workspace name validation.

Please do not repeat comments that have already been made and rejected.

File: fiftyone/core/odm/workspace.py
Comment (benjaminpkane) [inline fiftyone/core/odm/workspace.py:89]: Nice!

File: fiftyone/core/dataset.py
Comment (swheaton) [reply fiftyone/core/dataset.py:3927]: If there's a slug clash with an existing workspace, `_validate_workspace_name()` will raise an error.

**Comment by swheaton** [comment]
@benjaminpkane do you approve of this PR? If there's more needed for UI or URLs, let's add it in a follow on PR.

**Review by benjaminpkane — APPROVED** [review]
Yes, LGTM

**Pull request merged**
